habitat/core/dataset.py

```diff
@@ -101,3 +105,140 @@ def default(self, object):
 
     def from_json(self, json_str: str) -> None:
         raise NotImplementedError
+
+    def filter_episodes(
+        self, filter_fn: Callable[[Episode], bool]
+    ) -> "Dataset":
+        """
+        Returns a new dataset with only the filtered episodes from the original
+        dataset.
+        Args:
+            filter_fn: Function used to filter the episodes.
+        Returns:
+            The new dataset.
+        """
+        new_episodes = []
+        for episode in self.episodes:
+            if filter_fn(episode):
+                new_episodes.append(episode)
+        new_dataset = copy.copy(self)
+        new_dataset.episodes = new_episodes
+        return new_dataset
+
+    def get_splits(
+        self,
+        num_splits: int,
+        max_episodes_per_split: Optional[int] = None,
+        remove_unused_episodes: bool = False,
+        collate_scene_ids: bool = True,
+    ) -> List["Dataset"]:
+        """
+        Returns a list of new datasets, each with a subset of the original
+        episodes. All splits will have the same number of episodes.
+        Args:
+            num_splits: The number of splits to create.
+            max_episodes_per_split: If provided, each split will have up to
+                this many episodes. If it is not provided, each dataset will
+                have len(original_dataset.episodes) // num_splits episodes. If
+                max_episodes_per_split is provided and is larger than this
+                value, it will be capped to this value.
+            remove_unused_episodes: Once the splits are created, the extra
+                episodes will be destroyed from the original dataset. This
+                saves memory for large datasets.
+            collate_scene_ids: If true, episodes with the same scene id are
+                next to each other. This saves on overhead of switching between
+                scenes, but means multiple sequential episodes will be related
+                to each other because they will be in the same scene.
+        Returns:
+            A list of new datasets, each with their own subset of episodes.
+        """
+
+        assert (
+            len(self.episodes) >= num_splits
+        ), "Not enough episodes to create this many splits."
+
+        new_datasets = []
+        if max_episodes_per_split is None:
+            max_episodes_per_split = len(self.episodes) // num_splits
+        max_episodes_per_split = min(
+            max_episodes_per_split, (len(self.episodes) // num_splits)
+        )
+        if num_splits * max_episodes_per_split == len(self.episodes):
```

Comment:
Why do we have this special case? Does it implements same logic as non exact size match regarding randomization, collating scenes?

Reply:
It is faster to use slices rather than make new lists, but the else branch can work for exact splits as well. I can remove it if you want.